tfe/data_source_oauth_client.go

```diff
@@ -3,23 +3,67 @@ package tfe
 import (
 	"context"
 	"fmt"
+	"time"
 
 	"github.com/hashicorp/go-tfe"
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
 )
 
 func dataSourceTFEOAuthClient() *schema.Resource {
 	return &schema.Resource{
 		Read: dataSourceTFEOAuthClientRead,
 		Schema: map[string]*schema.Schema{
 			"oauth_client_id": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				AtLeastOneOf: []string{"oauth_client_id", "name", "service_provider"},
+			},
+			"organization": {
 				Type:     schema.TypeString,
-				Required: true,
+				Optional: true,
+			},
+			"name": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				RequiredWith: []string{"organization"},
+			},
+			"service_provider": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				RequiredWith: []string{"organization"},
+				ValidateDiagFunc: validation.ToDiagFunc(validation.StringInSlice(
+					[]string{
+						string(tfe.ServiceProviderAzureDevOpsServer),
+						string(tfe.ServiceProviderAzureDevOpsServices),
+						string(tfe.ServiceProviderBitbucket),
+						string(tfe.ServiceProviderBitbucketServer),
+						string(tfe.ServiceProviderBitbucketServerLegacy),
+						string(tfe.ServiceProviderGithub),
+						string(tfe.ServiceProviderGithubEE),
+						string(tfe.ServiceProviderGitlab),
+						string(tfe.ServiceProviderGitlabCE),
+						string(tfe.ServiceProviderGitlabEE),
+					},
+					false,
+				)),
+			},
+			"service_provider_display_name": {
+				Type:     schema.TypeString,
+				Computed: true,
 			},
 			"api_url": {
 				Type:     schema.TypeString,
 				Computed: true,
 			},
+			"callback_url": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"created_at": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
```

Comment:
In case it isn't clear, this is an attribute returned by the API. I added all attributes exposed by `tfe.OAuthClient` to the data source schema. I don't think we have an immediate use for that information, but it could be useful for distinguishing OAuth Client instances.

Reply:
We don't normally include every possible field (most resources omit created/updated-at), but if you find this useful I think that's OK to leave it here. If anyone else has a strong opinion on this please let me know!